### PR TITLE
ProtobufList

### LINKED ISSUE #16436: Output all rows as single list in Protobuf format
Currently `Protobuf` outputs each row as a separate message, but some consumers require a list.
We could add a new format `ProtobufList` that does that.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduce format `ProtobufList`. Fixes #16436

Introduce IO format "ProtobufList" with protobuf schema

    // schemafile.proto
    message Envelope {
      message MessageType {
        uint32 colA = 1;
        string colB = 2;
      }
      repeated MessageType mt = 1;
    }

where "Envelope" is a hard-coded/expected top-level message and
"MessageType" is a message with user-provided name containing the table
fields to export/import, e.g.

    SELECT * FROM db1.tab1 FORMAT ProtobufList SETTINGS format_schema =
    'schemafile:MessageType'

As a result, the new format wraps a list of messages (one per row) into
a single, containing message. Compare that to the schema of the existing
IO formats "Protobuf" and "ProtobufSingle":

    message MessageType {
      uint32 colA = 1;
      string colB = 2;
    }

The new format does not save space compared to the existing formats, but
it is conceptually a bit more beautiful and also more convenenient.

Implementation details:

- Created new files ProtobufList(Input|Output)Format which use the
  existing ProtobufSerializer mechanism. The goal was to reuse as much
  code as possible and avoid copypasta.

- I was torn between inheriting from I(Input|Output)Format vs.
  IRow(Input|Output)Format for ProtobufList(Input|Output)Format. The
  former is chunk-based which can be better for performance. Since the
  ProtobufSerializer mechanism is row-based but data is generally passed
  around in chunks, I decided for the latter to leverage the existing
  chunk <--> row mapping code in IRow(InputOutput)Format.

- A new ProtobufSerializer called ProtobufSerializerEnvelope was
  introduced (--> ProtobufSerializer.cpp). It represents the top-level
  message which encloses the list of inner nested messages, i.e. the
  rows.

- With the new format, parsing the schema file and matching the fields in
  the schema file to table column works like for the old formats. The only
  difference is that parsing starts one level below the "Envelope" (-->
  ProtobufSchema.cpp). This is more natural than forcing customers to
  have table columns start with "Envelope".

- Creation of the ProtobufSerializer tree also works like before. What
  is different is that we finally add a ProtobufSerializerEnvelope as
  new root of the tree. It's only purpose is to write/read the top-level
  message for the first/last row to write/read.

Caveats:

- The low-level serialization code in ProtobufWriter uses an internal
  buffer which is flushed to the output file only in endMessage().
  In the existing "Protobuf" format, this happens once per row, in the
  new format this happens only at the end of the serialization
  since row-level messages now call start/endNestedMessage(). As a
  future TODO to, the buffer should be flushed also in
  start/endNestedMessage() to reduce memory consumption.

### CODE REVIEW

**Comment by rschu1ze** [comment]
> ** Still contains a bug during import which I am trying to fix **

That bug is fixed in the latest push. I'll work on some tests based on the existing Protobuf tests.

File: src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp [hunk 1]
```diff
@@ -95,9 +95,7 @@ void registerOutputFormatParquet(FormatFactory & factory)
 namespace DB
 {
 class FormatFactory;
-void registerOutputFormatParquet(FormatFactory &)
-{
-}
+void registerOutputFormatParquet(FormatFactory &) {}
```
Comment (Avogar) [inline src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp:98]: Unrelated change

File: src/Formats/ProtobufSerializer.cpp [hunk 2]
```diff
@@ -2648,13 +2750,19 @@ namespace
...
-
-#if 0
-            LOG_INFO(&Poco::Logger::get("ProtobufSerializer"), "Serialization tree:\n{}", get_root_desc_function(0));
-#endif
-
-            return message_serializer;
+            if (!with_envelope)
+            {
+                *root_serializer_ptr = message_serializer.get();
+                LOG_INFO(&Poco::Logger::get("ProtobufSerializer"), "Serialization tree:\n{}", get_root_desc_function(0));
```
Comment (Avogar) [inline src/Formats/ProtobufSerializer.cpp:2756]: This log was under `#if 0` (because it's used only for debugging). And now there are some failed tests in CI because of this unexpected output.

File: src/Formats/ProtobufSchemas.cpp [hunk 3]
```diff
@@ -33,22 +35,36 @@ class ProtobufSchemas::ImporterWithSourceTree : public google::protobuf::compile
 
     const google::protobuf::Descriptor * import(const String & schema_path, const String & message_name)
     {
-        // Search the message type among already imported ones.
-        const auto * descriptor = importer.pool()->FindMessageTypeByName(message_name);
-        if (descriptor)
-            return descriptor;
```
Comment (Avogar) [inline src/Formats/ProtobufSchemas.cpp:39]: Why did you remove it?

File: src/Formats/ProtobufSerializer.cpp (hunk 2 above)
Comment (rschu1ze) [reply src/Formats/ProtobufSerializer.cpp:2756]: You are right ... I should have ifdef'd that away before pushing. Fixed.

File: src/Formats/ProtobufSchemas.cpp (hunk 3 above)
Comment (rschu1ze) [reply src/Formats/ProtobufSchemas.cpp:39]: The previous logic first probed a cache for the descriptor of an already imported schema (l. 37-39). If it could not be found in the cache, the import was done manually (l. 41-42).

However, importer.Import() probes the same cache as well (*). This means the manual probing was redundant.

(*)

Comment (Avogar) [reply src/Formats/ProtobufSchemas.cpp:39]: Not really, here we search in cache by message name and in importer.Import() we search in cache by schema path. It prevents cases when we try to work with different schemas with same message names. I built ClickHouse server from this PR and made some tests, I was able to get segfault on server stop because of this change:
File simple_schema.proto:
```
syntax = "proto3";

 message MessageType {
    uint32 colA = 1;
    string colB = 2;
}
```
File simple_schema2.proto is the copy of simple_schema.proto. 

```bash
$ clickhouse-client -q "select number as colA, 'String' as colB from numbers(10) format Protobuf settings format_schema='simple_schema:MessageType'" > data.bin
$ clickhouse-client -q "select * from file('data.bin', 'Protobuf', 'colA UInt32, colB String') settings format_schema='simple_schema:MessageType'" > /dev/null
$ clickhouse-client -q "select * from file('data.bin', 'Protobuf', 'colA UInt32, colB String') settings format_schema='simple_schema2:MessageType'"
Received exception from server (version 22.3.1):
Code: 434. DB::Exception: Received from localhost:9000. DB::Exception: Cannot parse 'simple_schema2.proto' file, found an error at line 3, column 11, "MessageType.colA" is already defined in file "simple_schema.proto".: While executing File. (CANNOT_PARSE_PROTOBUF_SCHEMA)
(query: select * from file('data.bin', 'Protobuf', 'colA UInt32, colB String') settings format_schema='simple_schema2:MessageType')
```

And if I try to stop ClickHouse server I will see this:
```
[libprotobuf FATAL ../contrib/protobuf/src/google/protobuf/descriptor.cc:810] CHECK failed: checkpoints_.empty():
Segmentation fault (core dumped)
```

File: src/Formats/ProtobufSchemas.h [hunk 4]
```diff
@@ -28,14 +28,36 @@ class FormatSchemaInfo;
...
+    {
+        // Return descriptor for a top-level message with a user-provided name.
+        // Example: In protobuf schema
+        //   message MessageType {
+        //     string colA = 1;
+        //     int32 colB = 2;
+        //   }
+        // message_name = "MessageType" returns a descriptor. Used by IO
+        // formats Protobuf and ProtobufSingle.
+        No,
```
Comment (rschu1ze) [inline src/Formats/ProtobufSchemas.h:41]: I am wondering if it is okay in CH to use scoped enums for boolean parameters? They are a bit more readable in method invocations on the callers end.

Example:

`func(true, false) // or: func(/*with_envelope = */ true, /*with_length = */ false);`

vs.

`func(WithEnvelope::Yes, WithLength::No);`

Comment (Avogar) [reply src/Formats/ProtobufSchemas.h:41]: It's ok to use it

File: src/Formats/ProtobufSchemas.cpp (hunk 3 above)
Comment (rschu1ze) [reply src/Formats/ProtobufSchemas.cpp:39]: That is a very good point. The semantics were changed and I will roll my change back to get the original behavior. Thanks!

I tried to reproduce the error/segfault with your repro but unfortunately failed. (More specifically, both SELECT * queries aborted with an error that the schemafile could not be found. I thought the user_files/ path may be misconfigured and provided a custom config file to clickhouse-server (via -C) to configure user_files but that somehow made the server access data in /var/lib/clickhouse/ instead of data in my home directory. On Mac, /var/ can unfortunately not be made permissive easily ... But perhaps I just need to gain more experience with the configuration since I am pretty new to the project.)

On a related note, I suppose that (generally speaking) cache lookups may in some cases be counterintuitive. E.g., a user could wonder why modifications of a schemafile are not getting picked up by ClickHouse which simply continues to use the old version from the cache found by the message name. Unfortunately, I also could not create a scenario which shows that because somehow all queries in my local setup are fully executed inside clickhouse-client (instead of clickhouse-server) so that the cache is wiped between queries.

**Comment by Avogar** [comment]
@rschu1ze will you add tests or you need help with it?

**Comment by rschu1ze** [comment]
Yes, and I am currently looking into ClickHouse's test frame framework + how to extend the existing tests for protobuf.

**Comment by KochetovNicolai** [comment]
Sorry, I did not push the button until now, so the checks would start now.
Feel free to ask any questions about our test infra

**Comment by rschu1ze** [comment]
With respect to the end-to-end (SQL) tests in tests/queries/0_stateless: Tests consist of a shell script (which sometimes calls a Python script) and a reference file, e.g.
- 00825_protobuf_format_persons.sh
- 00825_protobuf_format_persons.reference

The reference file contains the expected output. I wonder by which mechanism the actual and expected outputs  are checked against each other? The shell scripts implement no check, so there must be something else and the [docs](https://clickhouse.com/docs/en/development/adding_test_queries/) are not really clear on this one.

**Comment by Avogar** [comment]
We have a special tool for running tests named `clickhouse-test` (https://github.com/ClickHouse/ClickHouse/blob/master/tests/clickhouse-test). So, you can run `clickhouse-test <your_test_name>` and it will run your test and will compare output with reference file. For example to run 00825_protobuf_format_persons.sh test from `0_stateless` folder we can run the following:
```
$ ../../clickhouse-test 00825_protobuf_format_persons
```
To use `clickhouse-test` you need to have running ClickHouse server, by default it tries to connect to localhost and use default ports. You can run `clickhouse-test --help` for more details.

**Comment by Avogar** [comment]
There are some problems with building ClickHouse in CI, most likely it's because updated contribs in this PR (I guess you updated them accidentally in the last commit). We should delete these changes from this PR, if you don't know how, I can help you.

**Comment by rschu1ze** [comment]
Uh, neither _git clean --force -d_ nor _git reset --hard_ are able to delete the updated contribs. Any other magic tricks would be appreciated :-)

**Comment by rschu1ze** [comment]
Nevermind, _git submodule deinit -f ._ and _git submodule update --init_ did the trick.

**Comment by rschu1ze** [comment]
Hm, I rebuild the change locally like you suggested (starting from the previous commit which was the merge + adding the new test manually) and it contained no changed contribs. But for some reason they are still there after pushing. I will check again.

**Comment by Avogar** [comment]
Let me try to help (if I will be able to push to your branch)

**Comment by Avogar** [comment]
> Let me try to help (if I will be able to push to your branch)

`permission denied` :(
But I see you handle it by yourself, great.

**Comment by rschu1ze** [comment]
Next time, I'll be more careful to not push tainted commits in the first place :)

**Comment by KochetovNicolai** [comment]
Just in case: this pr breaks some builds. And style test is broken.
Scroll down all the action statuses, and you would see the relevant checks (`ClickHouse special build check (actions)` and `Style Check (actions)`). Description ion the details, [e.g.](https://s3.amazonaws.com/clickhouse-test-reports/35152/08d3f7fa14319484410de572b1da8974dbec1089/style_check__actions_/typos_output.txt)

**Comment by rschu1ze** [comment]
Fixed, tnx.
(I am positively surprised that CI build results come in rather quickly. I know of other systems where a developer would sometimes wait for 24+ hours to get all results.)

**Review by Avogar — APPROVED** [review]

**Comment by Avogar** [comment]
Test failures are not related to changes
